### PR TITLE
Template API

### PR DESCRIPTION
**Pull request by buehlefs**
PR containing work on the template API

- [x] Implement database changes
- [x] Implement all GET endpoints
- [x] Implement create endpoint for Tabs
- [x] Implement update endpoints
- [x] Implement delete endpoints
- [x] Implement (and kick off) background tasks that bind plugins to tabs based on their plugin filters (must be run for new plugins and for changes in the plugin filters)

### CODE REVIEW

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 1]
```diff
@@ -12,21 +12,65 @@
...
+    plugin_filter: dict = json.loads(filter_string)
+    plugins = DB.session.query(RAMP).all()
+
+    def get_plugins_from_filter(filter_dict: dict) -> list[RAMP]:
+        # TODO: add match for plugin description
+        match filter_dict:
+            case {"and": filter_expr}:
+                return set.intersection(*map(set, [get_plugins_from_filter(f) for f in filter_expr]))
+            case {"or": filter_expr}:
+                return set.union(*map(set, [get_plugins_from_filter(f) for f in filter_expr]))
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:48]:
```suggestion
                return set.union(*[set(get_plugins_from_filter(f)) for f in filter_expr]))
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 2]
```diff
@@ -12,21 +12,65 @@
...
+            case {"not": filter_expr}:
+                return set.difference(set(plugins), set(get_plugins_from_filter(filter_expr)))
+            case {"tag": tag}:
+                return [plugin for plugin in plugins if tag in [t.tag for t in plugin.tags]]
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:52]: Do not nest list expressions. (And where possible prefer generator expressions `(x for x in y)` (round brackets) over list expressions as list expressions always create a full list object in memory.

```suggestion
                has_tag = lambda p,t: any(tag.tag == t for tag in p.tags)
                return [plugin for plugin in plugins if has_tag(plugin,tag)]
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 3]
```diff
@@ -12,21 +12,65 @@
...
+            case {"version": version}:
+                # TODO: match version with semver
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:54]: See <https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/c82222c6948f18fa6cb638ded0e4d12824f318f6/qhana_plugin_registry/db/filters.py#L107> for inspiration on how this might be implemented

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 4]
```diff
@@ -12,21 +12,65 @@
...
+                return [plugin for plugin in plugins if version == plugin.version]
+            case {"name": name}:
+                return [plugin for plugin in plugins if name == plugin.name]
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:57]: You can directly construct sets using comparisons:
```suggestion
                return {plugin for plugin in plugins if name == plugin.name}
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 5]
```diff
@@ -12,21 +12,65 @@
...
+            case _:
+                return []
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:59]: Return an empty set:
```suggestion
                return set()
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 6]
```diff
@@ -12,21 +12,65 @@
...
+def evaluate_plugin_filter(filter_string: str) -> list[RAMP]:
+    """Get a list of plugins from a filter.
+
+    Args:
+        plugin_filter (str): The filter string.
+
+    Returns:
+        list: A list of plugins.
+    """
+    # TODO: store filter as dict in DB to avoid parsing
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:38]: Change the field to the JSON type if you want SQLAlchemy to do the (un)parsing for you. This is a change to the database and requires a new database revision!
Alternatively add a new property/function to the template tab that returns the parsed json.

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 7]
```diff
@@ -12,21 +12,65 @@
...
+    plugin_filter: dict = json.loads(filter_string)
+    plugins = DB.session.query(RAMP).all()
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:40]: This may grow large over time. Either implement filtering the plugins in batches (can be large batches ~500) or use other techniques to avoid loading too many data from DB at once. (Not high prio for this PR so just adding an issue for later + TODO comment here is sufficient for now.)

Suggestion:
Use a dict mapping plugin.id -> plugin. Then only use sets containing plugin IDs for the rest of this method (create a set from the dict keys for a set of all plugins). This allows you to remove the unsafe hash method from the plugin class and also avoids the equality test there entirely. The filters can then be entirely represented with set operations (at least `and` and `or`) and the leaf nodes create integer sets. (`not` may require special handling to get this correct...)

Note that set operations can also be used like this:
```python
# set difference
{1,2,3} - {1,2}  # {3}
# union
{1,2} | {2,3}  # {1,2,3}
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 8]
```diff
@@ -12,21 +12,65 @@
...
+
+    return get_plugins_from_filter(plugin_filter)
+
+
 @CELERY.task(name=f"{_name}.apply_filter_for_tab", bind=True, ignore_result=True)
 def apply_filter_for_tab(self, tab_id):
-    # TODO: evaluate plugin filter and set tab plugins accordingly
-    pass
+    q = select(TemplateTab).where(TemplateTab.id == tab_id)
+    found_tab: Optional[TemplateTab] = DB.session.execute(q).scalar_one_or_none()
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:67]: This functionality is implemented by the `get_by_id` method of the `IdMixin` class (but the types can only be correctly expressed with the self type of python 11...).

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 9]
```diff
@@ -37,34 +37,39 @@ def evaluate_plugin_filter(filter_string: str) -> list[RAMP]:
...
-                return set.intersection(*map(set, [get_plugins_from_filter(f) for f in filter_expr]))
+                return set.intersection(
+                    *(get_plugins_from_filter(f) for f in filter_expr)
+                )
             case {"or": filter_expr}:
-                return set.union(*map(set, [get_plugins_from_filter(f) for f in filter_expr]))
+                return set.union(*(get_plugins_from_filter(f) for f in filter_expr))
             case {"not": filter_expr}:
-                return set.difference(set(plugins), set(get_plugins_from_filter(filter_expr)))
+                return set(plugin_mapping.keys()) - get_plugins_from_filter(filter_expr)
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:53]: If I remember right you can directly use the keys view of a dict for set operations:
```suggestion
                return plugin_mapping.keys() - get_plugins_from_filter(filter_expr)
```

File: pyproject.toml [hunk 10]
```diff
@@ -7,7 +7,7 @@ license = "Apache-2.0"
 readme = "README.md"
 
 [tool.poetry.dependencies]
-python = "^3.8"
+python = "^3.10"
```
[the remark this answers is not in the document]
Comment (infacc) [reply pyproject.toml:10]: Fixed in 5f05a5f58f8e0f5dc430a5f71c87914400a9dd08

File: qhana_plugin_registry/templates/licenses.html [hunk 11]
```diff
@@ -221,12 +221,12 @@
...
+        <td>UNKNOWN</td>
```
[the remark this answers is not in the document]
Comment (infacc) [reply qhana_plugin_registry/templates/licenses.html]: This probably happened due to a different black version installed on my system. The current black version used in this project is [`22.8.0`](https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/096aefa30e93369c429c1546084dc29e9b04cff7/docs/requirements.txt#L9) (for which the correct license is picked).

Fixed in 096aefa30e93369c429c1546084dc29e9b04cff7.

File: qhana_plugin_registry/tasks/plugin_filter.py (hunk 7 above)
Comment (infacc) [reply qhana_plugin_registry/tasks/plugin_filter.py]: Batch filtering done in 506daa746689ae7790736a2ea9732c5333882933.

File: qhana_plugin_registry/tasks/plugin_filter.py (hunk 6 above)
Comment (buehlefs) [reply qhana_plugin_registry/tasks/plugin_filter.py]: Validating the filter certainly would not hurt. If the filter is not validated, then an incorrect filter must not crash the following steps.

File: qhana_plugin_registry/api/models/templates.py [hunk 12]
```diff
@@ -54,7 +54,7 @@ class TemplateTabSchema(ApiObjectSchema):
     description = ma.fields.String(required=True, allow_none=False)
     location = ma.fields.String(required=True, allow_none=False, validate=Length(max=255))
     sort_key = ma.fields.Integer(required=True, allow_none=False, default=0)
-    plugin_filter = ma.fields.String(required=True, allow_none=False)
+    filter_string = ma.fields.String(required=True, allow_none=False)
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:57]: You could use a custom field to handle the validation like the Version field in this example: <https://marshmallow.readthedocs.io/en/stable/examples.html#validating-package-json>

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 13]
```diff
@@ -29,8 +28,48 @@
...
+                if Version(p.version) in specifier
+            }
+        case {"name": name}:
+            return {p_id for p_id, p in plugin_mapping.items() if name == p.name}
+        case _:
+            TASK_LOGGER.warning(f"Invalid filter: '{filter_dict}'")
+            return set()
+
+
+def evaluate_plugin_filter(plugin_filter: dict) -> Generator[RAMP, None, None]:
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:72]: Prefer the more general Iterator type over the specific Generator type if the intention of the Generator is only to be an iterator.

**Review by buehlefs — COMMENTED** [review]
Only some minor comments.

File: qhana_plugin_registry/api/models/templates.py (hunk 12 above)
Comment (infacc) [reply qhana_plugin_registry/api/models/templates.py]: I ended up using the `@validates` decorator to validate the `filter_string` field (3622b0496d1bb855c5854211397ddddcd0e4741d).

File: qhana_plugin_registry/api/models/templates.py [hunk 14]
```diff
@@ -54,9 +56,45 @@ class TemplateTabSchema(ApiObjectSchema):
...
+            case {"and": l} | {"or": l}:
+                if not isinstance(l, list):
+                    raise ma.ValidationError("Invalid plugin filter.")
+                for f in l:
+                    TemplateTabSchema.validate_filter(f)
+            case {"not": f}:
+                TemplateTabSchema.validate_filter(f)
+            case {"name": f} | {"tag": f}:
+                if not isinstance(f, str):
+                    raise ma.ValidationError("Invalid plugin filter.")
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:74]: If you test for different errors, then try to give the users meaningful information what went wrong.
At least state that a "tag" or "name" condition was not a string (maybe include the whole condition in the error message or just the unexpected value).
Additionally, you could keep track of the path to the error and include that in the error message.
(This is relevant for all error messages in this method.)

File: qhana_plugin_registry/api/models/templates.py [hunk 15]
```diff
@@ -60,39 +60,65 @@ class TemplateTabSchema(ApiObjectSchema):
     plugins = ma.fields.Nested(ApiLinkSchema)
 
     @staticmethod
-    def validate_filter(filter_dict: dict):
+    def validate_filter(filter_dict: dict, path: list[str] = []):
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:63]: do not use mutable default values for parameters! (they are only evaluated once when the function is defined)
```suggestion
    def validate_filter(filter_dict: dict, path: tuple[str, ...] = ()):
```
or
```suggestion
    def validate_filter(filter_dict: dict, path: str = ""):
```

File: qhana_plugin_registry/api/models/templates.py [hunk 16]
```diff
@@ -60,39 +60,65 @@ class TemplateTabSchema(ApiObjectSchema):
...
+        if len(filter_dict) > 1:
+            raise ma.ValidationError(
+                [f"Invalid plugin filter: Only one filter key allowed per level."] + path
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:66]:
```suggestion
                f"Invalid plugin filter: Only one filter key allowed per level. (Path: {path})"
```

File: qhana_plugin_registry/api/models/templates.py [hunk 17]
```diff
@@ -60,39 +60,65 @@ class TemplateTabSchema(ApiObjectSchema):
...
+            )
+        k, v = next(iter(filter_dict.items()))
+        current_path = path + [f"{k}: {v}"]
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:69]:
```suggestion
        key = next(iter(filter_dict.keys()))
        current_path = path + f".{k}"
```

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 18]
```diff
@@ -44,10 +44,14 @@ def get_plugins_from_filter(filter_dict: dict, plugin_mapping: dict) -> set[RAMP
 
     match filter_dict:
         case {"and": filter_expr}:
+            if not filter_expr:
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:47]: You may want to write unit tests for the plugin filter functions to test the edge cases properly.

File: tests/plugin_filter/util.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    plugin_exists = tmp_db.session.query(
+        exists().where(RAMP.name == name, RAMP.version == version)
+    ).scalar()
+    if plugin_exists:
+        plugin = (
+            tmp_db.session.query(RAMP)
+            .filter(RAMP.name == name, RAMP.version == version)
+            .first()
+        )
+        plugin.plugin_id = f"{name}@{version}"
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:90]: name and version of an existing plugin should never change in the database. In fact, your filters ensure that name and version are already the values you assign here. Just remove the lines that cannot have an effect

File: tests/plugin_filter/util.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    tmp_db.session.commit()
+    plugin_id = (
+        tmp_db.session.query(RAMP.id)
+        .filter(RAMP.name == name, RAMP.version == version)
+        .first()[0]
+    )
+    return plugin_id
+
+
+def create_plugins(
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:119]: With property based testing you need to carefully consider what should be randomized and what should be fixed. Randomization creates more test cases (and can find more errors) but also limits the possible assertions you can make in the test case (which can hide errors). Something like this, that creates some plugins in the DB may be better as a known list of plugins implemented as a pytest fixture for some tests.

File: tests/plugin_filter/util.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+            TemplateTab.sort_key == sort_key,
+            TemplateTab.filter_string == filter_string,
+        )
+        .first()[0]
+    )
+    apply_filter_for_tab.apply(args=(template_tab_id,))
+    return template_tab_id
+
+
+def create_tags(tmp_db, tags: list[str] | None = None) -> list[PluginTag]:
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:173]: Use <https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/feature/template-api/qhana_plugin_registry/db/models/plugins.py#L236> instead

File: tests/plugin_filter/util.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+        tag_exists = tmp_db.session.query(
+            exists().where(PluginTag.tag == tag.tag)
+        ).scalar()
+        if not tag_exists:
+            tmp_db.session.add(tag)
+    tmp_db.session.commit()
+    return tmp_db.session.query(PluginTag).filter(PluginTag.tag.in_(tags)).all()
+
+
+def compare_plugins(plugin: RAMP, other: RAMP):
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:189]: For most tests it should suffice to just check that the database id (`plugin.id`) is the same.

File: tests/plugin_filter/util.py [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+    assert plugin.name == other.name
+    assert plugin.description == other.description
+    assert plugin.version == other.version
+    assert plugin.plugin_id == other.plugin_id
+    assert len(plugin.tags) == len(other.tags)
+    assert all(tag in other.tags for tag in plugin.tags)
+
+
+def filter_matches_plugin(filter_dict: dict, plugin: RAMP) -> bool:
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:198]: The best tests you can do with this is to let hypothesis generate a bunch of plugins, then a filter, then assert that every plugin that would be added to a template tab with this filter satisfies this method (and all others don't).

File: tests/plugin_filter/util.py [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+import random
+from hypothesis import strategies as st
+from packaging.specifiers import Specifier
+import json
+from sqlalchemy.sql import exists
+
+
+def filter_strategy():
+    return st.one_of(
+        st.fixed_dictionaries({"tag": st.text()}),
```
[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/util.py:27]: Currently, the empty string for names/tags is handled just like every other name/tag. Hypothesis creates plugins with empty names and an "empty tag", which match the filters `{"name": ""}` and `{"tag": ""}` respectively.

[the remark this answers is not in the document]
Comment (buehlefs) [reply tests/plugin_filter/util.py:27]: The database may not prevent this, but I sure hope we prevent empty names in the API models. If we do, this test can assume non-empty names. If we don't, please open an issue so that we do in the future...

[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/util.py:27]: The empty string as plugin name is not allowed. However, the empty string as tag is allowed, I.e. if the plugin runner provides a plugin that has the empty string as one of its tags, the QHAna UI shows an empty tag.

I'll assume plugin names of length `>= 1` in the tests.

[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/util.py:27]: The empty string as plugin name is not allowed. However, the empty string as tag is allowed, I.e. if the plugin runner provides a plugin that has the empty string as one of its tags, the QHAna UI shows an empty tag.

I'll assume plugin names of length `>= 1` in the tests.

[the remark this answers is not in the document]
Comment (buehlefs) [reply tests/plugin_filter/util.py:27]: You can assume the same thing for tags. I guess we can just ignore an empty tag in plugins without throwing an error (but we should not add empty tags to the database). Can you create an issue for that (that empty tags get added to the DB)?

[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/util.py:27]: I set `min_size=1` for tag names and created an issue (#14).

File: pyproject.toml [hunk 25]
```diff
@@ -46,6 +45,9 @@ PyMySQL = ["PyMySQL"]
 [tool.poetry.scripts]
 #flask = 'flask.cli:main'
 
+[tool.poetry.group.dev.dependencies]
```
Comment (buehlefs) [inline pyproject.toml:48]: Merge this with the existing dev dependencies key (line 28). If there is a poetry update that changes the key, then we may need to look into that in more detail.

File: tests/plugin_filter/test_plugin_filter.py [hunk 26]
```diff
@@ -14,143 +14,316 @@
...
+
+    # test single name filter
+    tab_plugins = update_plugin_filter(tmp_db, client, template_tab, {"name": name})
+    filtered_plugin_ids = {
+        p_id for p_id, in tmp_db.session.query(RAMP.id).filter(RAMP.name == name).all()
+    }
+    assert len(tab_plugins) > 0
+    assert len(tab_plugins) == len(filtered_plugin_ids)
+    for plugin in tab_plugins:
+        assert plugin.id in filtered_plugin_ids
```
Comment (buehlefs) [inline tests/plugin_filter/test_plugin_filter.py:123]: you could also convert `tab_plugins` to a plugin id set and then directly compare the sets in a single assert. This would be easier to read and probably also gives better error messages.

File: pyproject.toml (hunk 25 above)
Comment (infacc) [reply pyproject.toml]: According to the [note about the dev-dependencies section](https://python-poetry.org/blog/announcing-poetry-1.2.0/#dependency-groups) the dev-dependencies key is changed from `tool.poetry.dev-dependencies` to `tool.poetry.group.dev.dependencies` and migrating to the new group syntax is recommended.

> As the dev-dependencies is now deprecated, projects should migrate to the new group syntax as soon as possible. Keep in mind that the group syntax is a new feature of Poetry 1.2, and your project will not be buildable with Poetry 1.1 after migrating.

Comment (buehlefs) [reply pyproject.toml]: Then merge all into the new syntax and add a note in the readme that poetry >= 1.2 is required

File: tests/plugin_filter/test_plugin_filter.py [hunk 27]
```diff
@@ -113,30 +113,26 @@ def test_plugin_filter_name(tmp_db, client, template_tab, plugins, name: str):
...
+    tab_plugin_ids = update_plugin_filter(tmp_db, client, template_tab, {"name": name})
     filtered_plugin_ids = {
         p_id for p_id, in tmp_db.session.query(RAMP.id).filter(RAMP.name == name).all()
     }
-    assert len(tab_plugins) > 0
-    assert len(tab_plugins) == len(filtered_plugin_ids)
-    for plugin in tab_plugins:
-        assert plugin.id in filtered_plugin_ids
+    assert len(tab_plugin_ids) > 0
+    assert tab_plugin_ids == filtered_plugin_ids
```
Comment (buehlefs) [inline tests/plugin_filter/test_plugin_filter.py:121]: one last thing: add a helpful message to the asserts, e.g., "filtering by a single name failed"
The messages can either say what has failed (but do not describe just the assert statement...) or where the failure can be expected (only if you have a rough idea from the detected error already)

File: tests/plugin_filter/util.py [hunk 28]
```diff
@@ -111,12 +113,13 @@ def update_plugin_filter(tmp_db, client, template_tab: TemplateTab, filter_dict:
...
     assert response.status_code == 200
     apply_filter_for_tab.apply(args=(template_tab.id,))
-    return (
+    plugins = (
         tmp_db.session.query(TemplateTab)
         .filter(TemplateTab.id == template_tab.id)
         .first()
         .plugins
     )
+    return set(plugin.id for plugin in plugins)
```
Comment (buehlefs) [inline tests/plugin_filter/util.py:122]: directly use a set comparison:

```suggestion
    return {plugin.id for plugin in plugins}
```

File: pyproject.toml (hunk 25 above)
Comment (infacc) [reply pyproject.toml]: I migrated to poetry version `>=1.2` in the `main` branch (0ae547fab1b3698596ebac1cc0ba65c3fd269eb7).

Comment (buehlefs) [reply pyproject.toml]: Then please resolve any open merge conflict. We can merge this branch after that and add the changes to the API (adding new links etc.) in a new PR.

**Review by buehlefs — COMMENTED** [review]
I would say, that we should merge this PR now. New changes and bugfixes can happen in additional PRs.
